### ISSUE TITLE
Feature: Login with LinkedIn

### ISSUE THREAD

**Issue by robotdan** [FusionAuth/fusionauth-issues#34]
### Problem
The problem is I can't login to FusionAuth with LinkedIn. :-) 

### Solution
Add the ability to login with LinkedIn.

### Alternatives/workarounds
The LinkedIn OAuth2 option is not OpenID Connect compliant, but you can hack it.

Use the OpenId Connect IdP configuration in FusionAuth. Currently this method has a limitation of only being able to request the email address so additional claims cannot be merged into the FusionAuth user.

You must use the following endpoints in your configuration and the following lambda to extract the email address.

Because LinkedIn requires two API calls to get email + profile we have to choose one or the other in the OpenID Connect configuration. 

#### OpenId Connect configuration
**Client authentication method**
`Request body (client_secret_post)`
**Authorization endpoint**
`https://www.linkedin.com/oauth/v2/authorization`
**Token endpoint**
`https://www.linkedin.com/oauth/v2/accessToken`
**Userinfo endpoint**
`https://api.linkedin.com/v2/emailAddress?q=members&projection=(elements*(handle~))`
**Scope**
`r_emailaddress`

##### Prior to version 1.17.0

Use an OpenID Connect reconcile Lambda to set the email address. 

```
function reconcile(user, registration, jwt) {
  user.email = jwt.elements[0]['handle~'].emailAddress;
}
```

##### Version >= 1.17.3

Beginning with version 1.17.3, you can specify the `emailClaim` on the OpenID connect configuration. 

This will default to `email`, and can be modified for other use cases. For LinkedIn, you will use the following value:  `elements/0/handle~0/emailAddress` as the `Email claim`.  Note the `~0` is not a typo, this is the JSON Pointer escape syntax for a `~`. 

### Update on OIDC Configuration

In FusionAuth version 1.21.0 we enabled PKCE for all OpenID Connect clients. The OIDC specification indicates that an IdP should ignore the code_challenge and code_challenge_method if they are not supported. 

> RFC 7636
>
> *5 Compatibility*
> As the OAuth 2.0 [RFC6749] server responses are unchanged by this
specification, client implementations of this specification do not
need to know if the server has implemented this specification or not
and SHOULD send the additional parameters as defined in Section 4 to
all servers.

Of course, Microsoft / LinkedIn do not do anything to spec - and they break if you send these values to the normal authorize endpoint `https://www.linkedin.com/oauth/v2/authorization`.

You will see an error like this in the Event Log. 

```
Request to the [https://www.linkedin.com/oauth/v2/accessToken] endpoint failed. Status code [403].

Error response is 
{
  "error" : "access_denied",
  "error_description" : "Not enough permissions to access Native PKCE protocol"
}
```
I believe this is due to their PKCE requirements. See https://stackoverflow.com/a/64999805/3892636. 

According to this document from Microsoft, you have to request PKCE be enabled for your LinkedIn application, and once this is complete, use a different authorization endpoint. `https://www.linkedin.com/oauth/native-pkce/authorization`

https://docs.microsoft.com/en-us/linkedin/shared/authentication/authorization-code-flow-native

### Additional context

https://docs.microsoft.com/en-us/linkedin/shared/authentication/authorization-code-flow?context=linkedin/context

### How to vote
Please give us a thumbs up or thumbs down as a reaction to help us prioritize this feature. Feel free to comment if you have a particular need or comment on how this feature should work.

**Comment by joydeb28** [comment]
Can you please update the LinkedIn changes also?

**Comment by robotdan** [comment]
@joydeb28 did you try the above configuration?

**Comment by joydeb28** [comment]
Yes. I am getting "A request to the OpenID Connect Token API has failed. Unable to complete this login request." at the time of login.

**Comment by robotdan** [comment]
If you enable debug on the OpenID Connect IdP configuration what is in the debug event log? (System > Event Log)

**Comment by joydeb28** [comment]
Event Logs:
Request to the [https://www.linkedin.com/oauth/v2/accessToken] endpoint failed. Status code [400]

Error response is 
{
  "error" : "invalid_request",
  "error_description" : "A required parameter \"client_secret\" is missing"
}

**Comment by robotdan** [comment]
I think LinkedIn is looking for the `client_secret` in the request body. 
https://docs.microsoft.com/en-us/linkedin/shared/authentication/authorization-code-flow?context=linkedin/consumer/context

So in your OpenID Connect IdP config, change Client Authentication method from "HTTP basic authentication" to "Request body".

**Comment by joydeb28** [comment]
I changed the Client authentication method to "client_secret_post" and it worked. I got the token.
But getting the below error.
{
  "elements" : [ {
    "handle~" : {
      "emailAddress" : "joydebmondal28@gmail.com"
    },
    "handle" : "urn:li:emailAddress:655618093"
  } ]
}
6/16/2020 06:31:03 PM GMT The identity provider was unable to reconcile the email address. An email address is required to complete this request.

**Comment by joydeb28** [comment]
I added the reconcile lambda as you mentioned. But the above error I am getting now.

**Comment by robotdan** [comment]
Dump the contents of the incoming `jwt` to see what is in it, and then adjust your lambda accordingly. 

Example:

```
function reconcile(user, registration, jwt) {
  console.debug(jwt, null, 2);
  //user.email = jwt.elements[0]['handle~'].emailAddress;
}
```

It could be the example Lambda above is not longer correct.

**Comment by joydeb28** [comment]
In the Event logs, I am getting only this.
```
{
  "elements" : [ {
    "handle~" : {
      "emailAddress" : "joydeb28@gmail.com"
    },
    "handle" : "urn:li:emailAddress:655618093"
  } ]
}
```

**Comment by robotdan** [comment]
Ok, so the lambda looks correct. This "hack" used to work, I think a change we made in 1.17.0 may have broken that. 

If I am correct, it will work on 1.16.0, but not 1.17.0. 

We'd either have to fully support a LinkedIn login option, or add another configuration option to allow a reconcile lambda to override the `user.email` field.

**Comment by joydeb28** [comment]
For now any suggestion to make it ready?

**Comment by robotdan** [comment]
No work around at the moment.

We've discuss removing this limitation on the lambda, and allowing it to set the `user.email` field and just add a warning to the Lambda configuration indicating the risks.

**Comment by robotdan** [comment]
Added an issue to review allowing the `email` claim to be modified by a lambda which would allow this use case to work. https://github.com/FusionAuth/fusionauth-issues/issues/691

**Comment by robotdan** [comment]
This should work for you again in version 1.17.2

**Comment by joydeb28** [comment]
Thanks for this quick fix.

**Comment by robotdan** [comment]
@joydeb28 let me know if it works for you, there may be an edge case yet where this may not work as you'd like.

**Comment by joydeb28** [comment]
@robotdan 
Not still the same problem.

**Comment by robotdan** [comment]
@joydeb28 once version 1.17.3 is available, you should be able to make LinkedIn work again. I updated the comments for the workaround in the main description above. 

The configuration will have a new field for Email claim which will default to `email. For LinkedIn you'll be able to set it to `elements/0/handle~0/emailAddress`.

**Comment by joydeb28** [comment]
Thanks @robotdan 
It's working, now I am getting the email.
I set `scope=r_liteprofile%20r_emailaddress%20w_member_social` but I'm not getting the profile details.

**Comment by joydeb28** [comment]
@robotdan 
Can you please help me to get the full profile?

**Comment by robotdan** [comment]
You'll have to use the Microsoft documentation to identify the correct scope or URL to use. 
https://docs.microsoft.com/en-us/linkedin/shared/authentication/authorization-code-flow?context=linkedin/consumer/context

**Comment by LohithBlaze** [comment]
@robotdan I'm trying to add login with LinkedIn, the above settings will only query for email address since the userinfo endpoint is https://api.linkedin.com/v2/emailAddress?q=members&projection=(elements*(handle~)) and not for profile information such as first name and last name but for profile information the userinfo endpoint should be set as https://api.linkedin.com/v2/me , how to basically query for endpoint which will return both the profile information and email address. The above configuration only adds email address.

https://docs.microsoft.com/en-us/linkedin/shared/integrations/people/profile-api

**Comment by LohithBlaze** [comment]
Hi @robotdan did you get a chance to review this

**Comment by robotdan** [comment]
It's happening... :-)
